### PR TITLE
Fix extract filter.

### PR DESCRIPTION
**Pull request by Shylock-Hg**
## What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### Issue(s) number: 
Close https://github.com/vesoft-inc/nebula-ent/issues/1551

#### Description:
It caused by extract filter don't reset ok status when the first logical and expr can be split.

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

**Review by jievince — DISMISSED** [review]

**Comment by czpmango** [comment]
Please add some unit tests that can override the modified behavior in `ExtractFilterVisitorTest.cpp`.

**Comment by czpmango** [comment]
I didn't reproduce the crash.
```
(root@nebula) [nba]> MATCH (v:player{name: 'Tim Duncan'})-[:like]->(t) WHERE ( (1 == 1 AND (is_edge(t))) OR (v.player.name == 'Tim Duncan')) RETURN v.player.name
+---------------+
| v.player.name |
+---------------+
| "Tim Duncan"  |
| "Tim Duncan"  |
+---------------+
Got 2 rows (time spent 7134/7445 us)
```

**Comment by Shylock-Hg** [comment]
> I didn't reproduce the crash.
> 
> ```
> (root@nebula) [nba]> MATCH (v:player{name: 'Tim Duncan'})-[:like]->(t) WHERE ( (1 == 1 AND (is_edge(t))) OR (v.player.name == 'Tim Duncan')) RETURN v.player.name
> +---------------+
> | v.player.name |
> +---------------+
> | "Tim Duncan"  |
> | "Tim Duncan"  |
> +---------------+
> Got 2 rows (time spent 7134/7445 us)
> ```

ent version

**Comment by czpmango** [comment]
> > I didn't reproduce the crash.
> > ```
> > (root@nebula) [nba]> MATCH (v:player{name: 'Tim Duncan'})-[:like]->(t) WHERE ( (1 == 1 AND (is_edge(t))) OR (v.player.name == 'Tim Duncan')) RETURN v.player.name
> > +---------------+
> > | v.player.name |
> > +---------------+
> > | "Tim Duncan"  |
> > | "Tim Duncan"  |
> > +---------------+
> > Got 2 rows (time spent 7134/7445 us)
> > ```
> 
> ent version

Can this issue be reproduced in the community edition? Bcz you are fixing a visitor rather than a rule that only the enterprise version has. For example, PushFilterDownGetNbrsRule also invoked this visitor.

**Comment by Shylock-Hg** [comment]
> > > I didn't reproduce the crash.
> > > ```
> > > (root@nebula) [nba]> MATCH (v:player{name: 'Tim Duncan'})-[:like]->(t) WHERE ( (1 == 1 AND (is_edge(t))) OR (v.player.name == 'Tim Duncan')) RETURN v.player.name
> > > +---------------+
> > > | v.player.name |
> > > +---------------+
> > > | "Tim Duncan"  |
> > > | "Tim Duncan"  |
> > > +---------------+
> > > Got 2 rows (time spent 7134/7445 us)
> > > ```
> > 
> > 
> > ent version
> 
> Can this issue be reproduced in the community edition? Bcz you are fixing a visitor rather than a rule that only the enterprise version has. For example, PushFilterDownGetNbrsRule also invoked this visitor.

Not sure, I guess it will.

File: src/graph/visitor/ExtractFilterExprVisitor.cpp [hunk 1]
```diff
@@ -276,6 +276,7 @@ bool ExtractFilterExprVisitor::visitLogicalOr(LogicalExpression *expr) {
         if (canNotPushedIndex != -1 || !canBeSplit) {
           return false;
         }
```
Comment (czpmango) [inline src/graph/visitor/ExtractFilterExprVisitor.cpp:278]: Crashed at `DCHECK(ok())` in ExprVisitorImpl.cpp:130.
Better to fix else branch:
```
 else {
      if (!canBePushed_) {
        return false;
      }
      operands[i]->accept(this);
    }
```
Alternatively, we could do some minor refactoring of the visitor, such as adding additional state variables other than "can" and "cannot" pairs to allow the internal logic to continue.

Comment (Shylock-Hg) [reply src/graph/visitor/ExtractFilterExprVisitor.cpp]: ok

File: src/graph/visitor/ExtractFilterExprVisitor.cpp [hunk 2]
```diff
@@ -276,6 +276,7 @@ bool ExtractFilterExprVisitor::visitLogicalOr(LogicalExpression *expr) {
...
+        canBePushed_ = true;  // reset when the LogicalAnd can be split
```
Comment (czpmango) [inline src/graph/visitor/ExtractFilterExprVisitor.cpp:279]: IMHO, this is not a good fix. Just fixed code path, but it's not a natural code logic.

File: tests/tck/features/bugfix/ExtractFilter.feature [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+  Background:
+    Given a graph with space named "nba"
+
+  Scenario: Extract filter bug
+    When executing query:
+      """
+      MATCH (v:player{name: 'Tim Duncan'})-[:like]->(t) WHERE ( (1 == 1 AND (NOT is_edge(t))) OR (v.player.name == 'Tim Duncan')) RETURN v.player.name
+      """
+    Then the result should be, in any order:
```
Comment (czpmango) [inline tests/tck/features/bugfix/ExtractFilter.feature:14]: Better to find the crash cases in community edition.

Comment (Shylock-Hg) [reply tests/tck/features/bugfix/ExtractFilter.feature:14]: Can't do that.

File: src/graph/visitor/ExtractFilterExprVisitor.cpp (hunk 2 above)
Comment (Shylock-Hg) [reply src/graph/visitor/ExtractFilterExprVisitor.cpp:279]: That's only what I can do.

File: src/graph/visitor/ExtractFilterExprVisitor.cpp (hunk 1 above)
Comment (Shylock-Hg) [reply src/graph/visitor/ExtractFilterExprVisitor.cpp:278]: Will lead to test fail.

**Review by czpmango — APPROVED** [review]
LGTM for now, but need to consider refactoring later.

**Review by Aiee — APPROVED** [review]

**Pull request merged**